r/R/trackplots.R

```diff
@@ -268,36 +277,79 @@ trackplot_combine <- function(tracks, side_plot = NULL, title = NULL, side_plot_
     assert_is(side_plot, "ggplot")
   }
 
+  # Calculate layout information on the plots
   heights <- list()
-  side_plot_row <- 1L
+  collapse_upper_margin <- rep.int(TRUE, length(tracks))
+  side_plot_row <- NULL
   areas <- NULL
+  last_region <- NULL
   for (i in seq_along(tracks)) {
     if (is(tracks[[i]], "trackplot")) {
-      if (tracks[[i]]$trackplot$takes_sideplot) {
+      if (tracks[[i]]$trackplot$takes_sideplot && is.null(side_plot_row)) {
         side_plot_row <- i
-      } else if (side_plot_row == i) {
-        side_plot_row <- side_plot_row + 1L
       }
-    }
-    heights <- c(heights, list(get_trackplot_height(tracks[[i]])))
-    if (i != length(tracks)) {
-      tracks[[i]] <- tracks[[i]] + 
-        ggplot2::guides(x="none") +
-        ggplot2::labs(x=NULL) +
-        ggplot2::theme(plot.margin=ggplot2::unit(c(0,0,0,0), "pt"))
+      # If we switch regions, don't collapse margins into the track above
+      if (!is.null(tracks[[i]]$trackplot$region)) {
+        if (!is.null(last_region) && last_region != tracks[[i]]$trackplot$region && i > 1) collapse_upper_margin[i] <- FALSE 
+        last_region <- tracks[[i]]$trackplot$region
+      }
+
+      # Preserve top and bottom margins if `keep_vertical_margin`
+      if (tracks[[i]]$trackplot$keep_vertical_margin) {
+        collapse_upper_margin[i] <- FALSE
+        if (i < length(tracks)) collapse_upper_margin[i+1] <- FALSE
+      }
```

Comment:
I wanted to keep the function so that plots are always shown top-to-bottom in the order given, so I didn't want to do any re-ordering in that way. My aim with this is to do as much collapsing as is possible while to keeping the order the same